airbyte-integrations/connectors/destination-iceberg/src/main/java/io/airbyte/integrations/destination/iceberg/IcebergConsumer.java

```diff
@@ -0,0 +1,222 @@
+        if (msg.getType() != Type.RECORD) {
+            return;
+        }
+        final AirbyteRecordMessage recordMessage = msg.getRecord();
+
+        // ignore other message types.
+        AirbyteStreamNameNamespacePair nameNamespacePair = AirbyteStreamNameNamespacePair.fromRecordMessage(
+            recordMessage);
+        WriteConfig writeConfig = writeConfigs.get(nameNamespacePair);
+        if (writeConfig == null) {
+            throw new IllegalArgumentException(String.format(
+                "Message contained record from a stream that was not in the catalog. namespace: %s , stream: %s",
+                recordMessage.getNamespace(),
+                recordMessage.getStream()));
+        }
+
+        // write data
+        Row row = new GenericRow(new Object[]{UUID.randomUUID().toString(), new Timestamp(recordMessage.getEmittedAt()),
+            Jsons.serialize(recordMessage.getData())});
+        boolean needInsert = writeConfig.addData(row);
+        if (needInsert) {
+            appendToTempTable(writeConfig);
+        }
+    }
+
+    private void appendToTempTable(WriteConfig writeConfig) {
+        String tableName = writeConfig.getFullTempTableName();
+        List<Row> rows = writeConfig.fetchDataCache();
+        //saveAsTable even if rows is empty, to ensure table is created.
+        // otherwise the table would be missing, and throws exception in close()
+        log.info("=> Flushing {} rows into {}", rows.size(), tableName);
+        spark.createDataFrame(rows, normalizationSchema).write()
+            // append data to temp table
+            .mode(SaveMode.Append)
+            //TODO compression config
+            .option("write-format", catalogConfig.getFormatConfig().getFormat().getFormatName()).saveAsTable(tableName);
+    }
+
+
+    /**
+     * call this method when receive a STATE AirbyteMessage ———— it is the last message
+     */
+    @Override
+    public void commit() throws Exception {
+    }
+
+    @Override
+    protected void close(boolean hasFailed) throws Exception {
+        log.info("close {}, hasFailed={}", this.getClass().getSimpleName(), hasFailed);
+        Catalog icebergCatalog = catalogConfig.genCatalog();
+        try {
+            if (!hasFailed) {
+                log.info("==> Migration finished with no explicit errors. Copying data from temp tables to permanent");
+                for (WriteConfig writeConfig : writeConfigs.values()) {
+                    appendToTempTable(writeConfig);
+                    String tempTableName = writeConfig.getFullTempTableName();
+                    String finalTableName = writeConfig.getFullTableName();
+                    log.info("=> Migration({}) data from {} to {}",
+                        writeConfig.isAppendMode() ? "append" : "overwrite",
+                        tempTableName,
+                        finalTableName);
+                    spark.sql("SELECT * FROM %s".formatted(tempTableName))
```

Comment:
Can you point me out to where the temp and full tables are created? is this done implicitly?

Reply:
here: 

https://github.com/airbytehq/airbyte/blob/b094eaccfe9c1c5dcb1fa1acd50d3112d6bb033d/airbyte-integrations/connectors/destination-iceberg/src/main/java/io/airbyte/integrations/destination/iceberg/IcebergConsumer.java#L135

Reply:
Imo temp tables should be created in the startTracked() method which is used for executing all the preparation steps before starting to receive records. It is better to fail fast if you can't create something than to find out midway while replicating data.

Reply:
Imo temp tables should be created in the startTracked() method which is used for executing all the preparation steps before starting to receive records. It is better to fail fast if you can't create something than to find out midway while replicating data.